File: go-practice/interview/ex02_singleton.go
package interview

import "sync"

// Singleton 使用 sync.Once 实现线程安全的单例
type Singleton struct {
	// TODO: 定义字段
}

var (
	once     sync.Once
	instance *Singleton
)

func GetInstance() *Singleton {
	// TODO: 返回单例实例
	return nil
}

func ResetInstance() {
	// 测试用：重置单例（实际生产不用）
	once = sync.Once{}
	instance = nil
}

File: java-practice/src/main/java/interview/Ex02Singleton.java
package interview;


/**
 * 单例模式
 * 实现线程安全的单例（双重检查锁 + 静态内部类两种方式）
 */
public class Ex02Singleton {

    // TODO: 在这里实现
    private static volatile Ex02Singleton instance;

    private Ex02Singleton() {}

    public static Ex02Singleton getInstance() {
        if(instance == null){
            synchronized(Ex02Singleton.class) {
                if(instance == null) {
                    instance = new Ex02Singleton();
                }
            }
        }
        return instance;
    }
}
